Implement osu! hidden mod support

Aergwyn opened the pull request:
I had to redo everything after I broke my branch but here is a "prototype" for HD mod in standard.

I'm not sure if this can change everything that needs to be changed to get it equal to stable but this should be a start that one can work with.

Especially unsure I'm of spinners because they changed a lot compared to stable.

EDIT: [Example Video](https://youtu.be/uBL3sFWF6I8)

- [x] adjust FadeIn and FadeOut to correct values
- [x] change SliderBody to fade out correctly
- [x] look into ApproachCircles not liking to be hurried
- [x] fix Spinners disappearing too early

Aergwyn committed:
add interface to adjust DrawableHitObjects

Aergwyn committed:
add ability to hide ApproachCircles

Aergwyn committed:
add ability to change FadeIn/FadeOut speed

Aergwyn committed:
moved exposed fields to right classes + added gradually fading slider

Aergwyn committed:
fixed spinner expiring too soon

Aergwyn committed:
fixed slider approach circles showing with mod active

Aergwyn committed:
added hiding spinner "details"

fixed initial slider circles playing hit animation

Aergwyn committed:
added option to fade out earlier

Aergwyn committed:
some CI fixes

- removed unnecessary usings
- removed unnecessary parentheses
- removed unnecessary this
- renamed IApplicableToDrawableHitObject to IApplicableToDrawableHitObjects
- fixed potential InvalidCastException

Aergwyn committed:
updated .csproj because it forgot the rename

why didn'T VS pick that one up? o.O

Aergwyn committed:
let ApproachCircle fade out if were done

smoogipoo commented:
My initial concern is that you've removed the .Expire()s on every hitobject type, which will cause them to be iterated over even while they're invisible.

smoogipoo commented on osu.Game.Rulesets.Osu/Mods/OsuMod.cs:
```diff
@@ -26,10 +27,43 @@ public class OsuModEasy : ModEasy
     }
 
-    public class OsuModHidden : ModHidden
+    public class OsuModHidden : ModHidden, IApplicableToDrawableHitObjects
```
Am not really a fan of directly referencing the drawable representations of hit objects and changing their properties. I think better is for the drawable representations to handle their own fadeouts.

How about changing this to inherit from IApplicableToHitObject, and setting a field like `PreEmptFadeOut` on the hitobject?

Aergwyn replied:
So the idea is to move all new fields I exposed to the hitobject instead? I'll try once at home.

Aergwyn replied:
Mh. If I move all of them they'd have to be in `OsuHitObject` as I'd have to cast for some fields otherwise. That doesn't sound desirable.
But if I move them all to `OsuHitObject` a field named `ShowApproachCircle` wouldn't make much sense (or one of the other 3).
If I merge them together under a generic name it isn't really specific what that field would do.

So I'd like to keep it as it is. What do you think?

Aergwyn replied:
@smoogipoo I summon thee for an opinion. So I can continue over the weekend if needed.

smoogipoo replied:
I don't understand the reason for why it doesn't make sense to put it in the hitobject?

Aergwyn replied:
This was mostly about those 4 flags that disable or hide something. Some changes are quite specific to the respective type of HitObject. Take hiding the ApproachCircles as an example.
I wouldn't be able to move that one into HitCircle as I'd have to cast the HitObject each time to see if it actually is a HitCircle. Having it in HitObject would work but Spinners and Sliders (except for their InitialCircle) don't have ApproachCircles.
It wouldn't feel right to have them there. Unless I'm overthinking this...

smoogipoo commented on osu.Game.Rulesets.Osu/Mods/OsuMod.cs:
```diff
@@ -26,10 +27,43 @@ public class OsuModEasy : ModEasy
+                        slider.InitialCircle.PlayHitAnimation = false;
+                        slider.InitialCircle.FadeInSpeed = osuDrawable.FadeInSpeed;
+                        slider.InitialCircle.FadeOutSpeed = osuDrawable.FadeOutSpeed;
+                        slider.InitialCircle.EarlyFadeOutTime = osuDrawable.EarlyFadeOutTime;
```
These should all be handled by the drawable hit objects.

smoogipoo commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs:
```diff
@@ -32,20 +38,30 @@ protected sealed override void UpdateState(ArmedState state)
+        protected virtual void UpdatePostState()
+        {
+            double duration = ((HitObject as IHasEndTime)?.EndTime ?? HitObject.StartTime) - HitObject.StartTime + EarlyFadeOutTime;
+            this.Delay(duration).Expire();
```
Isn't this going to break for non-hidden for hitcircles? Essentially `StartTime - StartTime + 0`, I think I would prefer that expire was explicitly done down each relevant code path in any case, in-case the transformations change at some point.

Aergwyn replied:
It works fine as I tested so far. It took me a while to understand how and why but it seems to be the `BeginDelayedSequence` that causes it to expire when all Transformations (and Delays) are completed. I didn't notice any cutoffs in normal play.

Aergwyn commented:
The `.Expire()`s are all handled in `UpdatePostState()` now.

Aergwyn committed:
Merge branch 'master' into add-standard-hd

Aergwyn committed:
Updated submodule osu-framework

Aergwyn committed:
moved fields to OsuHitObjects instead of their drawable representation

and other corrections (expiring per code path and not fading out ticks/repeats gradually)

Aergwyn committed:
fixed not being able to miss HitCircles with HD mod active

and
- made DrawableSpinner expire again
- made InitialCircle in DrawableSlider private again

peppy commented:
Talking with smoogi today, we came to the agreement that this stuff should be done directly on Drawable representations rather than adding extra items to the HitObject model classes. I think this means you may need to revert some of your last commits.

Aergwyn committed:
moved fields back to their drawable representation

Aergwyn committed:
Merge branch 'master' into add-standard-hd

Aergwyn committed:
Updated submodule osu-framework

peppy committed:
Merge branch 'master' into add-standard-hd

peppy committed:
Simplify iteration and casting code

peppy committed:
Remove need for visibility hacks

peppy committed:
Rename some variables

peppy commented:
Rather than FadeOutGradually on sliders, why not use the same `FadeOutSpeed` and `FadeOutPreempt`? The values can be based off the slider's duration.

Aergwyn committed:
removed unnecessary booleans to change sequences

also clamped "PreemptFadeOut" between 0 and 1000 ms.

Aergwyn committed:
readded accidentally removed using

VS pls ;A;

smoogipoo commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs:
```diff
@@ -14,6 +16,28 @@ public class DrawableOsuHitObject : DrawableHitObject<OsuHitObject>
+            }
+        }
+
+        public override bool IsPresent => base.IsPresent || State.Value == ArmedState.Idle && Time.Current >= HitObject.StartTime - TIME_PREEMPT;
```
This is pretty weird, why is this needed?

peppy replied:
to stop invisible circles from not receiving input

Aergwyn committed:
changed calculations of FadeIn and FadeOut to be more dynamic for AR

(AR is still static but that may come soon™)

Aergwyn committed:
Merge branch 'master' into add-standard-hd

Aergwyn committed:
Updated submodule osu-framework

Aergwyn committed:
minor QoL changes

- renamed HiddenMod to Hidden
- removed FadeIn/-Out multiplier affecting SliderTicks and RepeatPoints
- added some comments
- show a few more parts of the spinner so it doesn't feel/look dead

Aergwyn committed:
fix weirding out ApproachCircles

they don't like going fast

Aergwyn commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -88,50 +87,54 @@ protected override void UpdatePreemptState()
+            if (Hidden)
+                // TODO: figure out a smarter way to include the hit window so we are able to miss but don't keep
+                // it alive unnecessarily long.
+                this.FadeOut(TIME_PREEMPT * FadeOutSpeedMultiplier).Delay(ExpireAfter + 300).Expire();
```
I really don't like this `+ 300`. The `.Delay()` is trying to keep it alive so we can click on it before it expires. But that only works up to the `HitObject.StartTime`. Clicking early would work but without the addition you wouldn't be able to miss circles (300ms = hit window). Any ideas to do it better?

peppy replied:
if anything it should be referencing the hit window value directly. i think it was previously, right?

Aergwyn replied:
No, not in Lazer. Hit window values were referenced in stable snippets you gave me though. But at different locations.

Aergwyn replied:
Ah you can access the hit windows. So I could adress this without any constants.

Aergwyn committed:
get hit window from HitObject rather than using a constant

peppy commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs:
```diff
@@ -10,9 +10,23 @@ namespace osu.Game.Rulesets.Osu.Objects.Drawables
         public const float TIME_FADEIN = 400;
-        public const float TIME_FADEOUT = 500;
+
+        public bool Hidden;
```
no

Aergwyn replied:
Is it the name or the bool itself?
Should I go back to the 4 different ones I had before?

peppy replied:
can you list the four you are requiring? generally drawable hitobjects should not be aware of hidden at all (the mod should be doing all the moving)

Aergwyn replied:
I had
`ShowApproachCircle` and `PlayHitAnimation` for DrawableHitCircle,
`FadeOutGradually` for DrawableSlider and
`HideSpinnerDetails` for DrawableSpinner

peppy replied:
Let's go with those for now

peppy replied:
Please make sure to add xmldoc on each explaining what they do, will make it easier to follow for future adventurers

Aergwyn replied:
Going through it only 2 may be sufficient though. 🤔 
ShowApproachCircle and AllowEarlyFadeOut.
The latter would already describe what I need for all 3 types anyways. Also I need something in DrawableOsuHitObject too that allows me to use a different calculation for FadeIn (determined by a bool named `hidden` in stable).
Technically the description would be correct as the HitObjects do fade out earlier. Partially or completely.

peppy replied:
the different calculation should be done by the mod itself, not the drawable implementation

Aergwyn committed:
move fade in / duration calculations away from the DrawableHitobjects

mods can change it now if needed

Aergwyn committed:
use calculated FadeIn instead of constant when possible

Aergwyn committed:
remove unnecessary variable

peppy committed:
Merge branch 'master' into add-standard-hd

peppy committed:
Comment fixes

peppy committed:
Fix formatting

Aergwyn committed:
moved addition of HitWindow to DrawableHitCircle

no need for the Mod to know of this special case

peppy committed:
Merge remote-tracking branch 'upstream/master' into add-standard-hd

Aergwyn committed:
remove unnecessary check

Aergwyn commented:
Closing this as a better solution is in work and I realized this approach is as wrong as it could be in how fading out is done for the mod.
